### ISSUE TITLE
ty believes `tempfile.NamedTemporaryFile()` is not writable

### ISSUE THREAD

**Issue by bepri** [astral-sh/ty#2812]
### Summary

The generic `SupportsWrite[T]` protocol requires the following type signature:
```py
def write(self, s: T, /) -> object: ...
```
The return type of `NamedTemporaryFile()` has, among its many overloads:
```py
def write(self, s: AnyStr, /) -> int: ...
```
And yet, I get the following error from ty:
```
error[invalid-argument-type]: Argument to function `write_to` is incorrect
  --> example.py:11:14
   |
10 | with tempfile.NamedTemporaryFile() as file:
11 |     write_to(file, b"hello!")
   |              ^^^^ Expected `SupportsWrite[str | bytes]`, found `_TemporaryFileWrapper[bytes]`
   |
info: Function defined here
 --> example.py:7:5
  |
7 | def write_to(dest: SupportsWrite[AnyStr], msg: AnyStr) -> None:
  |     ^^^^^^^^ --------------------------- Parameter declared here
8 |     dest.write(msg)
  |
info: rule `invalid-argument-type` is enabled by default

Found 1 diagnostic
```

Here's the example code I used to trigger this error:
```py
import tempfile
from typing import AnyStr

from _typeshed import SupportsWrite


def write_to(dest: SupportsWrite[AnyStr], msg: AnyStr) -> None:
    dest.write(msg)

with tempfile.NamedTemporaryFile() as file:
    write_to(file, b"hello!")
```

I believe that since `NamedTemporaryFile()` returns an object capable of writing bytes (the default mode here is `w+b`), `write_to(file, b"hello!")` should pass. Pyright has no complaints about this block.

### Version

ty 0.0.17 (8cec85718 2026-02-13)

**Comment by carljm** [comment]
Thanks for the report!

Here's a minimal repro with no imports, so all types are visible:

```py
from typing import AnyStr, overload, Protocol, TypeVar

T_contra = TypeVar("T_contra", contravariant=True)

class MyProtocol(Protocol[T_contra]):
    def write(self, s: T_contra) -> object: ...

class BothWriter:
    @overload
    def write(self, s: str) -> int: ...
    @overload
    def write(self, s: bytes) -> int: ...
    def write(self, s: str | bytes) -> int: return 0

def write_to(dest: MyProtocol[AnyStr], msg: AnyStr) -> None: ...

write_to(BothWriter(), b"hello!")  # ERROR: Expected MyProtocol[str | bytes]
```

The root cause of this is #2799 -- we are unioning disjunctive constraints and solving the `AnyStr` typevar to `str | bytes` when we should be solving it to `bytes` for this call.

**Comment by lengau** [comment]
This appears to be fixed for me in ty 0.0.20

**Comment by carljm** [comment]
Yeah, looks like this was actually fixed by https://github.com/astral-sh/ruff/pull/23277 -- thanks @Hugo-Polloli!
